runtime/gc_glue_java/FlattenedArrayObjectScanner.hpp

```diff
@@ -0,0 +1,258 @@
+		, fomrobject_t *endPtr
+		, uintptr_t elementSize
+		, uintptr_t elementStride
+		, uintptr_t *descriptionBasePtr
+#if defined(OMR_GC_LEAF_BITS)
+		, uintptr_t *leafBasePtr
+#endif /* defined(OMR_GC_LEAF_BITS) */
+		, uintptr_t flags)
+	: GC_HeadlessMixedObjectScanner(env, scanPtr, elementSize, flags | GC_ObjectScanner::indexableObject)
+	, _env(env)
+	, _elementSizeWithoutPadding(elementSize)
+	, _descriptionBasePtr(descriptionBasePtr)
+#if defined(OMR_GC_LEAF_BITS)
+	, _leafBasePtr(leafBasePtr)
+#endif /* defined(OMR_GC_LEAF_BITS) */
+	/* Pass 0 for scanMap, as the indexable iterator does not use the scanMap */
+	, _indexableScanner(env, arrayPtr, basePtr, limitPtr, scanPtr, endPtr, 0, elementStride, flags)
+	{
+		_typeId = __FUNCTION__;
+	}
+
+	MMINLINE void
+	initialize(MM_EnvironmentBase *env)
+	{
+#if defined(OMR_GC_LEAF_BITS)
+		GC_HeadlessMixedObjectScanner::initialize(env, _descriptionBasePtr, _leafBasePtr);
+#else /* defined(OMR_GC_LEAF_BITS) */
+		GC_HeadlessMixedObjectScanner::initialize(env, _descriptionBasePtr);
+#endif /* defined(OMR_GC_LEAF_BITS) */
+
+		_indexableScanner.initialize(env);
+
+		/* The HeadlessMixedObjectScanner will setNoMoreSlots() causing us
+		 * to miss other elements of the array
+		 */
+		setMoreSlots();
+	}
+
+public:
+
+	/**
+	 * @param[in] env The scanning thread environment
+	 * @param[in] objectPtr pointer to the array to be processed
+	 * @param[in] allocSpace pointer to space within which the scanner should be instantiated (in-place)
+	 * @param[in] flags Scanning context flags
+	 * @param[in] splitAmount If >0, the number of elements to include for this scanner instance; if 0, include all elements
+	 * @param[in] startIndex The index of the first element to scan
+	 */
+	MMINLINE static GC_FlattenedArrayObjectScanner *
+	newInstance(MM_EnvironmentBase *env, omrobjectptr_t objectPtr, void *allocSpace, uintptr_t flags, uintptr_t splitAmount, uintptr_t startIndex = 0)
+	{
+		GC_FlattenedArrayObjectScanner *objectScanner = (GC_FlattenedArrayObjectScanner *)allocSpace;
+		GC_ArrayObjectModel *arrayObjectModel = &(env->getExtensions()->indexableObjectModel);
+		J9Class *clazzPtr = J9GC_J9OBJECT_CLAZZ(objectPtr, env);
+		J9ArrayClass *j9ArrayClass = (J9ArrayClass *) clazzPtr;
+
+		/* TODO are these always the same? */
+		Assert_MM_true(j9ArrayClass->componentType == j9ArrayClass->leafComponentType);
+
+		J9Class *elementClass = j9ArrayClass->componentType;
+		omrarrayptr_t arrayPtr = (omrarrayptr_t)objectPtr;
+		
+		uintptr_t sizeInElements = arrayObjectModel->getSizeInElements(arrayPtr);
+		uintptr_t elementSize = J9_VALUETYPE_FLATTENED_SIZE(elementClass);
+		uintptr_t elementStride = J9ARRAYCLASS_GET_STRIDE(clazzPtr);
+		fomrobject_t *basePtr = (fomrobject_t *)arrayObjectModel->getDataPointerForContiguous(arrayPtr);
+		fomrobject_t *limitPtr = (fomrobject_t *)((uintptr_t)basePtr + sizeInElements*elementStride);
+		fomrobject_t *scanPtr = (fomrobject_t *)((uintptr_t)basePtr + startIndex*elementStride);
```

Comment:
Unfortunately with compressed/full selection runtime we can not use `fomrobject_t` pointer math directly any more. Please use `GC_SlotObject::addToSlotAddress()` (in this case) instead. See examples in ObjectScanner.hpp and other scanners.

Reply:
I'm not sure if that is correct. This is adding the total size in bytes of the object, which should be correct no matter if we're compressed or not.

Reply:
Right. I put this comment wrong place when re-do it after refresh. I intended to put it for line 147. However I am not sure line 147 is correct - (splitAmount * elementStride) is in bytes but added to pointer. Anyway, any `fomrobject_t *` pointer math in this PR should support runtime compressed/full selection

Reply:
formatting: please add spaces around multiplication sign `*`

Reply:
Should I wrap it in brackets, like this?
```c++
(fomrobject_t *)((uintptr_t)basePtr + (startIndex * elementStride));
```

Reply:
Yes, please. I like it

Reply:
Fixed!